compiler/noirc_frontend/src/hir_def/types.rs

```diff
@@ -1588,6 +1607,47 @@ impl Type {
         }
     }
 
+    // If `self` and `expected` are function types, tries to coerce `self` to `expected`.
+    // Returns None if this produced an error, otherwise returns Some(make_error) meaning
+    // that regular unification must still be done (and it could produce an error).
+    pub fn try_fn_to_unconstrained_fn_coercion(
+        &self,
+        expected: &Type,
+        expression: ExprId,
+        interner: &mut NodeInterner,
+        errors: &mut Vec<TypeCheckError>,
+        make_error: impl FnOnce() -> TypeCheckError,
+    ) -> Option<impl FnOnce() -> TypeCheckError> {
+        // If `self` and `expected` are function types, `self` can be coerced to `expected`
+        // if `self` is unconstrained and `expected` is not. The other way around is an error, though.
+        let (
+            Type::Function(params, ret, env, unconstrained_self),
+            Type::Function(_, _, _, unconstrained_expected),
+        ) = (self.follow_bindings(), expected.follow_bindings())
+        else {
+            // No coercion needed
+            return Some(make_error);
+        };
+
+        // unconstrained status matches: no error
+        if unconstrained_self == unconstrained_expected {
+            return Some(make_error);
+        }
```

Comment:
The control-flow in this function seems very odd to me, returning the same `make_error` function or `None`. We don't need to return it at all since it is never changed.

We also don't need this check at all since it should be done within `unify_with_coercions` below.